rdcore/kargs: call `zipl` on s390x after editing BLS

jlebon opened the pull request:
Otherwise the changed kargs won't take effect.

jlebon commented on src/bin/rdcore/kargs.rs:
```diff
@@ -26,10 +28,14 @@ pub fn kargs(config: &KargsConfig) -> Result<()> {
         .context("visiting BLS options")?;
+
+        if changed && cfg!(target_arch = "s390x") {
+            runcmd!("zipl", "--target", mount.mountpoint())?;
```
Not 100% sure if that's all we need. /cc @nikita-dubrovskii

Basing this off of `ignition-firstboot-complete`, which just calls `zipl`, except that here the boot partition isn't necessarily mounted on `/boot` so we need to point `zipl` at it: https://github.com/coreos/ignition/blob/3885b9f07c96baf7dcc12c85b914ded151c8fd6e/systemd/ignition-firstboot-complete.service#L26

bgilbert replied:
Also, if a bare `zipl` works, I'm confused by all the explicit arguments in `install_bootloader()`.  If zipl indeed knows how to read BLS configs, could it do that in the `install_bootloader()` case too?

nikita-dubrovskii replied:
I'll check it.

nikita-dubrovskii replied:
So, `--target` option doesn't work without specifying `--image` and/or others. what could be used is `--blsdir` , but in such case `/boot` shoud be `rw` accessible.

jlebon replied:
> what could be used is `--blsdir` , but in such case `/boot` shoud be `rw` accessible. 

Huh interesting, does it actually need write access to `/boot`, or just for the boot partition to be mounted at `/boot`? If the former, what does it write there?

jlebon commented:
OK, updated this now! One thing that became obvious is that we need a little more sophistication here because I want `rdcore kargs` to work equally well whether it's the first boot or not. This meant having the s390x code auto-detect it and doing the right thing.

@nikita-dubrovskii Would you be able to test this on s390x? Ideally using both `--firstboot-kargs` and an Ignition config which uses Ignition kargs.

nikita-dubrovskii commented:
Thanks @jschintag!

jlebon commented:
OK weird, I really thought `zipl` was already in the initramfs for FCOS at this point. @nikita-dubrovskii Can you submit that patch to the f-c-c repo?

nikita-dubrovskii commented:
> OK weird, I really thought `zipl` was already in the initramfs for FCOS at this point. @nikita-dubrovskii Can you submit that patch to the f-c-c repo?

Done - https://github.com/coreos/fedora-coreos-config/pull/1146

nikita-dubrovskii commented:
LGTM

jlebon commented:
Updated now to drop the tempdir move! (Good catch!) And also added @nikita-dubrovskii's `zipl` dracut patch here.

Because I'm very wary of code I can't test (I did compile-test it at least by fudging with the conditionals), I would appreciate still a final test of this PR using `--firstboot-kargs` and also using Ignition kargs.

nikita-dubrovskii commented:
> I would appreciate still a final test of this PR using `--firstboot-kargs` and also using Ignition kargs.

done, works well

jlebon commented:
Awesome, thanks! Will let @bgilbert have a final look before we merge.

bgilbert commented on src/s390x/zipl.rs:
```diff
@@ -32,15 +32,11 @@ use tempfile::Builder;
 /// * `disk` - optional device path on which to run chreipl
 /// * `firstboot` - adds ignition.firstboot karg
 /// * `firstboot_kargs` - enables ignition firstboot and adds provided kargs
```
Stale comment.

bgilbert commented on src/s390x/zipl.rs:
```diff
@@ -86,9 +79,12 @@ fn run_zipl<P: AsRef<Path>>(
-        if let Some(kargs) = firstboot_kargs {
+        let firstboot_kargs = std::fs::read_to_string(firstboot_file)
+            .context("reading firstboot kargs from ignition.firstboot")?;
+        let firstboot_kargs = firstboot_kargs.trim();
```
The `ignition.firstboot` file [isn't that simple](https://github.com/coreos/coreos-installer/blob/d14fe5dbb5a864eee4d45070a6a7eb3943b54249/src/install.rs#L307), unfortunately.

jlebon replied:
Ahhh of course. There's no easy way to solve this generically, but at least we can special-case the format we expect. Fix added!

bgilbert commented on src/s390x/mod.rs:
```diff
@@ -17,5 +17,6 @@ pub mod zipl;
 
 pub use dasd::{dasd_try_get_sector_size, image_copy_s390x, prepare_dasd};
 pub use zipl::install_bootloader;
+pub use zipl::run_zipl;
```
The distinction between these two functions isn't immediately obvious from the names.  Can we rename one or both?

jlebon replied:
Added another commit that cleans things up a bit more.

bgilbert reviewed (commented):
LGTM generally, and this is a nice cleanup, but I think `--firstboot-kargs` will put some extra garbage in the kernel command line.

bgilbert commented on src/s390x/zipl.rs:
```diff
@@ -95,9 +86,51 @@ pub fn install_bootloader<P: AsRef<Path>>(
+        );
+        assert!(extract_firstboot_kargs("set ignition_network_kcmdline=\"\n\"").is_err());
+        assert!(extract_firstboot_kargs("set ignition_network_kcmdline=\"foo\nbar\"").is_err());
+        assert!(extract_firstboot_kargs("foobar").is_err());
```
Other things to possibly test:
- Valid `set` line followed by a newline
- Valid `set` line preceded by something else
- Valid `set` line followed by something else

jschintag commented:
Thanks for the work. It works on s390x.
